### ISSUE TITLE
Template Types that references other Template Types in constraints are sometimes not expanded correctly

### ISSUE THREAD

**Issue by vzz3** [vimeo/psalm#6547]
First of all: sorry for the title! It is terrible, but I was not able to find a short adequate description for my observation.
My problem is that psalm complains about type declarations that includes templated classes if the templated class contains at least two Template Types where the second references the first one. In the following example this occurs at the Template Type `Tdao` of the class `RecordCache`.

```php
interface StrongEntity {}

class Customer implements StrongEntity {}

/**
 * @template T
 */
interface StrongEntityDao {}

/**
 * @implements StrongEntityDao<Customer>
 */
class CustomerDao implements StrongEntityDao {}

/**
 * @template Tmodel of StrongEntity
 * @template Tdao of StrongEntityDao<Tmodel> 
 */
class RecordCache {
    /**
     * 
     * @param class-string<Tmodel> $yRecordClass
     * @param Tdao $yDao
     */
    public function __construct(string $yRecordClass, $yDao) {
    }
}

class Test {
    /**
     * ERROR: Extended template param Tdao of RecordCache<Customer, CustomerDao> 
     * expects type StrongEntityDao<Tmodel:RecordCache as StrongEntity>, type CustomerDao given
     *
     * @var RecordCache<Customer,CustomerDao> 
     */
    public $c;
}

```

However, this false positive error that occurs at the Type declaration of `Test::c`, which looks like psalm can not expand `Tdao` correctly, occurs only at type declarations. If psalm infer the type from code, like constructor calls, psalm expand `Tdao` exactly as expected. 

```php
$c = new RecordCache(Customer::class, new CustomerDao()); // psalm defines the type of $c as "RecordCache<Customer,CustomerDao>"
```

Full working code example: https://psalm.dev/r/9281fd157b

**Comment by weirdan** [comment]
In this particular case, doesn't it make sense to have `RecordCache` generic over entity type alone, e.g. https://psalm.dev/r/edca17c0ed ?

**Comment by vzz3** [comment]
You are totally right. Unfortunately I oversimplified my example. In my Project the RecordCache class is extended many times. Especially the DB insert and update Methods are overwritten many times because I need to use some special DAO methods that inserts additional date within a transaction. And this needs to access the DAO object. The problem is now that the interface `StrongEntityDao` does not container further specialised insert and update methods.  
https://psalm.dev/r/e321bbfa54

Therefor, I want to to define the `dao` member in a template variable. Interestingly it works just fine if I define a class CustomerCache that extends RecordCache and use that class:
https://psalm.dev/r/b6524d8cb5

However, it is not possible to do the same thing with a quickly inline implemented anonymous class since it is not possible to reference without an error from psalm.
https://psalm.dev/r/341ab2495d
